iec/src/mir/mod.rs

```diff
@@ -7,51 +7,151 @@
 use codespan::{FileId, Span};
 use specs::prelude::*;
 use specs_derive::Component;
+use std::collections::BTreeMap;
 
-#[derive(Debug, Clone, PartialEq, Component)]
-#[storage(HashMapStorage)]
-pub struct Configuration {
-    pub resources: Vec<Entity>,
+pub fn register(world: &mut World) {
+    world.register::<Name>();
+    world.register::<Location>();
+    world.register::<Scope>();
+    world.register::<ScopeRef>();
+    world.register::<HasType>();
+    world.register::<Type>();
+    world.register::<Function>();
 }
 
 #[derive(Debug, Clone, PartialEq, Component)]
 #[storage(VecStorage)]
-pub struct Resource {}
-
-#[derive(Debug, Clone, PartialEq, Component)]
-#[storage(VecStorage)]
-pub struct Program {}
-
-/// A control element for starting one or more programs.
-///
-/// IEC 61131 defines a task as:
-///
-/// > An execution control element which is capable of invoking, either on a
-/// > periodic basis or upon the occurrence of the rising edge of a specified
-/// > Boolean variable, the execution of a set of program organization units,
-/// > which can include programs and function blocks whose instances are
-/// > specified in the declaration of programs.
-#[derive(Debug, Clone, PartialEq, Component)]
-#[storage(VecStorage)]
-pub struct Task {
-    pub trigger: Trigger,
-}
-
-#[derive(Debug, Copy, Clone, PartialEq)]
-pub enum Trigger {
-    Periodic,
-    RisingEdge { variable: Entity },
+pub struct Function {
+    pub local_variables: Vec<Entity>,
+    pub return_type: Entity,
+    pub parameters: Vec<Entity>,
 }
 
 /// The name for a component.
-#[derive(Debug, Clone, PartialEq, Component)]
+#[derive(Debug, Clone, Component)]
 #[storage(VecStorage)]
 pub struct Name(pub String);
 
+impl<T> PartialEq<T> for Name
+where
+    str: PartialEq<T>,
```

Comment:
``` str: PartialEq<T>, ``` how does this work?

Reply:
This is the lazy way of saying "anything you can compare a `&str` with for equality, you can do with the `Name`".

If you think as where clauses more like "type-level equations" than constraints on a type variable. `T: PartialEq<str>` (T is comparable with a string) is just as valid as `str: PartialEq<T>` (strings are comparable with T).